Classy transfer_model

jlashner opened the pull request:
Hi Steven, here's my first shot at a classy transfer model. It's behavior is as follows:
 - If a 'class_obj' is specified in transfer params, it will attempt to use that class object instead of creating a new one. It will figure out on its own whether or not `.compute()` has been run by checking to see if the transfer function exists yet.
 - If no `class_obj` is specified, a new class object is created. It uses a few default parameters like `omega_b`, `omega_cdm`, and `h` based on the `cosmo` object, updated with any parameters passed via `class_params`.
 
With our modified class code, it is possible to obtain negative T(k), which will break this since you can't take `ln(T)`. If this transfer is only used to calculate the power spectrum, then it should be fine to return `ln(|T|)`, since P(k)~T(k)^2. However if other parts of your code use T(k) (not squared) then this might be a problem. I did a quick scan of hmf and couldn't really find T(k) being used outside of calculating P(k). Do you have an opinion on the best way to handle this? In any case, I added a `flip_T` transfer option which returns ln(|T|), but we should discuss if there is a better way to handle this.

steven-murray commented on hmf/density_field/transfer_models.py:
```diff
@@ -713,3 +718,65 @@ class EH(EH_BAO):
+            trans = self.class_obj.get_transfer(output_format='camb')
+            if not trans:
+                self.class_obj.compute()
+                trans = self.class_obj.get_transfer(output_format='camb')
```
I'm not very familiar with Class, but I feel like there might be a more elegant way to do this. Is there no other way to check if `compute` has been called other than to get the transfer and check if it's not None? Ideally, you'd have:

```
if not self.class_obj.condition:
    self.class_obj.compute()

trans = self.class_obj.get_transfer(output_format='camb')

steven-murray commented on hmf/density_field/transfer_models.py:
```diff
@@ -713,3 +718,65 @@ class EH(EH_BAO):
+        **class_object:**
+            Custom class object. If not set, a new Class object will be
+            created.
+        **flip_t:**
```
The transfer function in hmf is *only* used to get the power spectrum (at this point). I think it's reasonable just to always take the absolute for now. It might be good to post an issue about this though, and we can think about returning the actual transfer function rather than the log. I chose the log because at the time I thought it must always be positive, and the log is easier to take a spline of.

steven-murray replied:
Ah, you'll need to remove `flip_t` from the docsring as well :-)

steven-murray commented on hmf/density_field/transfer_models.py:
```diff
@@ -713,3 +718,65 @@ class EH(EH_BAO):
+                trans = self.class_obj.get_transfer(output_format='camb')
+
+            _k, _t = trans['k (h/Mpc)'], trans['-T_tot/k2']
+            if self.params['flip_T']:
```
Note point above about using absolute.

steven-murray commented on hmf/density_field/transfer_models.py:
```diff
@@ -20,6 +20,11 @@
+try:
+    import classy
+except ImportError as e:
+    pass
```
Set `HAVE_CLASS=False` here if it fails, then raise an error in the class `__init__` if it's false.

steven-murray commented on hmf/density_field/transfer_models.py:
```diff
@@ -713,3 +718,65 @@ class EH(EH_BAO):
+    _defaults = {"class_params": None, "class_obj": None,
+                 "flip_T": False}
+
+    def __init__(self, *args, **kwargs):
```
Here's the place to raise an error if `HAVE_CLASS=False`

steven-murray commented:
Also, before we merge, you should:

1. Add your name to the contributors list (actually, it might already be there)
2. Add your changes to the `CHANGELOG.rst`
3. Install `pre-commit` and run `pre-commit run -a` in the top-level of the repo. I think the Linter is failing because of some `black` formatting issues. This will fix those automatically (most likely).

We'll also have to resolve conflicts with master, but let's do those things first.

jlashner commented:
Great! Implemented all of your suggestions and changed a few things:
1. Added a test comparing CLASS and CAMB transfers with default params, making sure they are the same
2. I removed the `flip_T` argument, and if T is negative I'm printing a warning and just returning the absolute value. We're a bit concerned if this will work properly when splined, so  I think we still might want to implement an option where you don't have to take the log.
3. Had some flake8 problems about `\*` characters in the docstring I copied from the other transfer functions. I'm not sure why it didn't like those in the CLASS transfer but is ok with them in the other transfer functions, but I just removed them.

Let me know if there's anything else, thanks!

jlashner commented:
So I think this is failing my test because class is not installed on the test server. Do you have any idea how to add that? I think it needs to be manually installed. [Cobaya-install](https://cobaya.readthedocs.io/en/latest/installation_cosmo.html) can also be used to install a suite of cosmological packages which might be easier.

I did run the tests locally and they passed.

steven-murray commented on hmf/density_field/transfer_models.py:
```diff
@@ -713,3 +720,78 @@ class EH(EH_BAO):
+
+
+class CLASS(FromFile):
+    """
```
Add an `r` just before the quotes to be able to use the `\*\*` notation in the docstring (not that you have to do that).

steven-murray commented:
Hi @jlashner, did you progress with this? Can I assist in any way?

jlashner commented:
Hi Steven, sorry for leaving this out for so long. So I stopped working on this because we were running into some consistency issues between our modified CLASS model and the cosmology that's used throughout hmf, which is an astropy cosmology model. It seems to me that in order to be consistent we'd have to make a new cosmology class that wraps our CLASS object and maps all of the used astropy methods to CLASS methods. Do you have any input on whether this is the right thing to do, and if there would be any lingering astropy cosmology dependence in HMF? Thanks!

Anyway, after we figured out that we'd need to put more work into hmf for it to be consistent I decided to move ahead with other parts of the project while we looked into some other options or determined whether the CLASS wrapper was worth implementing in HMF. It looks like the only thing thats left is some minor docstring comments that I can implement. I was also holding off because the CLASS installation is not all that straightforward and I'm not sure the best way to do it through github actions. If you want to take a shot at it that would be super helpful!

steven-murray commented:
Hi @jlashner, thanks for the update!

How bad is the inconsistency? Is it a fundamental inconsistency between astropy and CLASS, or between your modification of CLASS only? Could we get away with using the code as-is for now, and making an issue that we'll want to someday update the CLASS wrapper?

The CLASS installation is indeed not straightforward. I think I can give it a go soon.
